### PR TITLE
Function-level similarity: CLI mode + function extraction

### PR DESCRIPTION
**Pull request by bzz**
Based on #155 and #157 

This is implemented part of the #150

TODO:
 - [x] CLI `--mode func|file` for function/file level ops
 - [x] hash: function extraction from UAST
 - [x] hash: default FE parameters for function/file level ops from ML team

Further improvements on hashing and query sides are tracked under #162

### CODE REVIEW

File: src/main/scala/tech/sourced/gemini/cmd/HashSparkApp.scala [hunk 1]
```diff
@@ -23,14 +23,15 @@ case class HashAppConfig(
...
+  verbose: Boolean = false,
+  mode: String = Gemini.fileSimilarityMode
 )
 
 /**
   * Apache Spark app that applied LSH to given repos, using source{d} Engine.
   */
 object HashSparkApp extends App with Logging {
-  val repoFormats = Seq("siva", "bare", "standard")
+  val repoFormats = Seq(Gemini.fileSimilarityMode,  Gemini.funcSimilarityMode)
```
Comment (carlosms) [inline src/main/scala/tech/sourced/gemini/cmd/HashSparkApp.scala]: Bug, right? `repoFormats` should not change, it's still used for `--format`, format of the stored repositories.

**Comment by carlosms** [comment]
I would maybe remove the `list_funcs.py`. If it's only for reference, having it in `src/main/python` with other code we do use will be confusing in the future.

Plus `extractFunctions` already has a link to the exact revision used as reference.

File: src/main/scala/tech/sourced/gemini/Gemini.scala [hunk 2]
```diff
@@ -27,7 +27,7 @@ class Gemini(session: SparkSession, log: Slf4jLogger, keyspace: String = Gemini.
     * @param limit
     * @param format
     */
-  def hash(reposPath: String, limit: Int = 0, format: String = "siva", docFreqPath: String = ""): Unit = {
+  def hash(reposPath: String, limit: Int = 0, format: String = "siva", mode: String, docFreqPath: String = ""): Unit = {
```
[the remark this answers is not in the document]
Comment (bzz) [reply src/main/scala/tech/sourced/gemini/Gemini.scala]: added in 89f2028

**Comment by bzz** [comment]
@carlosms all feedback addressed, PR description update with current scope, ready for another pass.

File: src/test/scala/tech/sourced/gemini/SparkHashSpec.scala [hunk 3]
```diff
@@ -36,11 +36,28 @@ class SparkHashSpec extends FlatSpec
...
+  "Hash" should "return correct number of functions" in {
+    val hashResult = hashWithNewGemini("func")
+    val funcs = hashResult.hashes
+    val fCount = funcs.count()
+    print(s"$fCount functions found")
+    funcs.select("doc").show(false)
+    fCount shouldEqual 54
+  }
+
+  "Hash" should "extract functions UASTs in func mode" in {
```
Comment (smacker) [inline src/test/scala/tech/sourced/gemini/SparkHashSpec.scala]: this test doesn't have any assert

**Pull request merged**